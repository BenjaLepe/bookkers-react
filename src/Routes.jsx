import React from 'react';
import { Route, Switch } from 'react-router-dom';
import NotFound from './views/NotFound';
import Login from './views/Login';
import BookDetail from './views/BookDetail';
import SignUp from './views/SignUp';
import Home from './views/Home';
import AuthContextProvider from './contexts/AuthContext';
import SessionProfile from './views/SessionProfile';
import SessionExpired from './views/SessionExpired';
import UserProfile from './views/UserProfile';
import CreateBook from './views/AddBook';
import EditUser from './views/EditUser';
import Navbar from './Components/Navbar';
import reviewReports from './views/Reports';
import EditBook from './views/EditBook';

export default function Routes() {
  return (
    <AuthContextProvider>
      <Navbar />
      <Switch>
        <Route exact path="/" component={Home} />
        <Route exact path="/login" component={Login} />
        <Route exact path="/sign-up" component={SignUp} />
        <Route exact path="/session-expired" component={SessionExpired} />
        <Route exact path="/session-profile" component={SessionProfile} />
        <Route exact path="/users/:id" component={UserProfile} />
        <Route exact path="/books/new" component={CreateBook} />
        <Route exact path="/books/:id" component={BookDetail} />
        <Route exact path="/edit-book/:id" component={EditBook} />
        <Route exact path="/edit-user/:id" component={EditUser} />
        <Route exact path="/books/:book_id/reviews/:review_id/reports" component={reviewReports} />
        <Route component={NotFound} />
      </Switch>
    </AuthContextProvider>
  );
}