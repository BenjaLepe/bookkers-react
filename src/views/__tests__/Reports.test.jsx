import React from 'react';
import renderer from 'react-test-renderer'
import { MemoryRouter } from 'react-router-dom';
import Routes from '../../Routes';

const { act } = renderer;

function TestRouter({ path }) {
  return (
    <MemoryRouter initialEntries={[ path ]}>
      <Routes />
    </MemoryRouter>
  );
}

const responseReview = {
  data: {
    data: {
      review: {
        UserId: 7,
        BookId: 7,
        title: 'testTitle',
        body: 'testBody',
        User: { username: 'testUsername' },
        likes: []
      },
      reports: [{
        id: 1,
        UserId: 8,
        User: { username: 'testUsername' },
        comment: 'testComment',
      }],
      totalPages: 1,
    },
  },
};

jest.mock('../../hooks/calls', () => ({
    getReviewReportsRoute: () => responseReview,
}));

describe('User Profile', () => {
  describe('When there is a normal user', () => {
    const user = {
      access_token: 'TestToken',
      token_type: 'Bearer',
      user: {
        id: 1,
        username: 'testUserName',
        first_name: 'testFirstName',
        last_name: 'testLastName',
        email: 'test@email.cl',
        description: 'testDescription',
      },
    };
    const sessionExp = new Date(
      new Date().getTime() + 1000 * 60 * 60 * 24, // One day in the future
    );
    const localStorageData = {
      user,
      sessionExp,
    };

    beforeEach(() => {
      global.Storage.prototype.getItem = jest.fn(
        (key) => JSON.stringify(localStorageData[key]),
      );
    });

    afterEach(() => {
      global.Storage.prototype.getItem.mockReset();
    });

    it('should redirect to not found page', async () => {
      let testRenderer;

      act(() => {
        testRenderer = renderer.create(<TestRouter path="/books/1/reviews/1/reports"/>);
      })

      await act(async () => {});

      const DOM = testRenderer.toJSON();
      expect(DOM).toMatchSnapshot();
    });
  });

  describe('When there is an admin user', () => {
    const user = {
      access_token: 'TestToken',
      token_type: 'Bearer',
      user: {
        id: 1,
        username: 'testUserName',
        first_name: 'testFirstName',
        last_name: 'testLastName',
        email: 'test@email.cl',
        description: 'testDescription',
        is_admin: true
      },
    };
    const sessionExp = new Date(
      new Date().getTime() + 1000 * 60 * 60 * 24, // One day in the future
    );
    const localStorageData = {
      user,
      sessionExp,
    };

    beforeEach(() => {
      global.Storage.prototype.getItem = jest.fn(
        (key) => JSON.stringify(localStorageData[key]),
      );
    });

    afterEach(() => {
      global.Storage.prototype.getItem.mockReset();
    });

    it('should renders the reports view', async () => {
      let testRenderer;

      act(() => {
        testRenderer = renderer.create(<TestRouter path="/books/1/reviews/1/reports"/>);
      })

      await act(async () => {});

      const DOM = testRenderer.toJSON();
      expect(DOM).toMatchSnapshot();
    });
  });
});