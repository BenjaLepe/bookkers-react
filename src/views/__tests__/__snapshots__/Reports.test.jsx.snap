// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Profile When there is a normal user should redirect to not found page 1`] = `
Array [
  <div
    className="nav-bg"
  >
    <div
      className="nav-bar container"
    >
      <div
        className="nav-group-logo"
      >
        <a
          aria-current="page"
          className="nav-link nav-title active"
          href="/"
          onClick={[Function]}
          style={Object {}}
        >
          Bookkers
        </a>
      </div>
      <div
        className="nav-group"
      >
        <a
          aria-current={null}
          className="nav-link"
          href="/session-profile"
          onClick={[Function]}
        >
          Profile
        </a>
        <a
          aria-current={null}
          className="nav-link"
          href="/books/new"
          onClick={[Function]}
        >
          Add book
        </a>
        <button
          className="nav-button"
          onClick={[Function]}
          type="button"
        >
          Log out
        </button>
      </div>
    </div>
  </div>,
  <div>
    <h2>
      404 Error
    </h2>
    <p>
      The page you are looking for does not exist
    </p>
  </div>,
]
`;

exports[`User Profile When there is an admin user should renders the reports view 1`] = `
Array [
  <div
    className="nav-bg"
  >
    <div
      className="nav-bar container"
    >
      <div
        className="nav-group-logo"
      >
        <a
          aria-current="page"
          className="nav-link nav-title active"
          href="/"
          onClick={[Function]}
          style={Object {}}
        >
          Bookkers
        </a>
      </div>
      <div
        className="nav-group"
      >
        <a
          aria-current={null}
          className="nav-link"
          href="/session-profile"
          onClick={[Function]}
        >
          Profile
        </a>
        <a
          aria-current={null}
          className="nav-link"
          href="/books/new"
          onClick={[Function]}
        >
          Add book
        </a>
        <button
          className="nav-button"
          onClick={[Function]}
          type="button"
        >
          Log out
        </button>
      </div>
    </div>
  </div>,
  <div>
    <div>
      <h2>
        Review
      </h2>
      <div
        className="box"
      >
        <div>
          <div>
            <a
              href="/users/7"
              onClick={[Function]}
            >
              @testUsername
                
            </a>
            <button
              onClick={[Function]}
              type="button"
            >
              Like
            </button>
            <button
              onClick={[Function]}
              type="button"
            >
              Delete
            </button>
            <a
              href="/books/7/reviews/undefined/reports"
              onClick={[Function]}
            >
              <button
                type="button"
              >
                Ver reportes
              </button>
            </a>
            <p>
              Total likes: 
              0
            </p>
          </div>
          <h3>
            testTitle
          </h3>
          <p>
            testBody
          </p>
          <div>
            <button
              className="delete"
              onClick={[Function]}
            >
              Report
            </button>
          </div>
        </div>
      </div>
    </div>
    <h2>
      Reports on this review
    </h2>
    <div
      className="box"
    >
      <div>
        <div>
          Autor del reporte:
          <a
            href="/users/8"
            onClick={[Function]}
          >
            testUsername
          </a>
        </div>
        <h3>
          Explicaci√≥n del reporte: 
          testComment
        </h3>
      </div>
    </div>
    <div
      className="page-set"
    >
      <button
        className="page-button"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        Back
      </button>
      <p
        className="pages"
      >
        1 : 1
      </p>
      <button
        className="page-button"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        Next
      </button>
    </div>
  </div>,
]
`;
